Feat: Filter pageviews to count created articles only

satrijo opened the pull request:
Closes #6534 

## What this PR does
This Pull Request modifies the main pageview statistics calculation. It filters the results to only include pageviews from articles newly created by program participants.

This changes the previous behavior, which included views from all edited articles (both new and existing).

## Screenshots
Before:
<img width="1169" height="404" alt="screenshot-2025-11-01_19-24-33" src="https://github.com/user-attachments/assets/a5477ee0-447d-4dec-a7d7-f1fa7df8cbdf" />


After:
<img width="941" height="891" alt="image" src="https://github.com/user-attachments/assets/b4f70432-38af-454e-bef5-debc1b551bf0" />

satrijo committed:
Feat: Add cached pageview breakdown to course stats

Implements the "created article pageviews" feature. Based on reviewer
feedback, this is now implemented as a mouseover tooltip on the main
"Article Views" stat, rather than a separate top-level metric.

This addresses all reviewer feedback:

1.  **Performance:**
    - Adds a new `view_sum_created` (bigint) column to the `courses`
      table via a database migration.
    - Updates `CourseCacheManager` to calculate and cache this new
      value during the background `update_cache_from_timeslices` task,
      following the same efficient pattern as `view_sum`.
    - The `Course#view_sum_created` method now reads this cached value,
      with a fallback to on-the-fly calculation for backward
      compatibility with older courses.

2.  **JSON API & Naming:**
    - `course.json.jbuilder` is simplified.
    - It now sends three pre-calculated, human-readable values:
      `view_count` (total), `view_count_created` (new), and
      `view_count_existing` (total - new).
    - Raw values (`_raw`) have been removed as requested.

3.  **Frontend:**
    - `OverviewStats.jsx` is simplified. It no longer does math
      or number formatting for the breakdown.
    - It just displays the pre-formatted values in the tooltip.

4.  **Specs:**
    - Adds 8 new backend specs to `course_cache_manager_spec.rb` and
      `course_spec.rb` to test the new caching logic and the
      fallback method.

Refactor view_sum_created calculation and JSON output

- Updated the Course model to improve the calculation of view_sum_created, ensuring it calculates on-the-fly for updated courses with new articles.
- Refactored CourseCacheManager to share the view_sum_created calculation logic, reducing code duplication.
- Enhanced the JSON output in course.json.jbuilder to conditionally include view_count_created and view_count_existing based on course update status and cached data validity.
- Updated locale descriptions for clarity on view count metrics.

Update course_spec.rb to include new_article_count in course update test

This change adds the new_article_count attribute to the course update method in the spec, ensuring that the test accurately reflects the latest model changes related to article views and caching logic.

Refactor: Clean up whitespace and mark calculate_view_sum_created_for_course as private

ragesoss commented:
We haven't had a significant amount of requests for this as a separate stat, so I'm not sure it makes sense to add it has a top-level stat in the UI like this. It might be more appropriate to add as additional detail that shows up upon mouseover, as we do with 'Commons Uploads' and 'Editors'.

satrijo commented:
> We haven't had a significant amount of requests for this as a separate stat, so I'm not sure it makes sense to add it has a top-level stat in the UI like this. It might be more appropriate to add as additional detail that shows up upon mouseover, as we do with 'Commons Uploads' and 'Editors'.

Hi, @ragesoss 

Thanks again for the excellent feedback. I've just pushed an update that implements your suggestion.
The main "Pageviews" stat now shows the original, combined total.

I've added the filtered "New Article Pageviews" count to a mouseover pop-up, following the pattern used by the "Editors" stat.
I've also updated the screenshots in the PR description to show the new behavior.

<img width="1204" height="399" alt="screenshot-2025-11-04_13-47-30" src="https://github.com/user-attachments/assets/6c6d0d0a-280f-4307-b0b2-11b7a1812cea" />


Could you please take another look when you have a chance?

Thanks!

gabina commented on app/models/course.rb:
```diff
@@ -614,6 +614,28 @@ def update_cache_from_timeslices
+          SQL
+        )
+      )
+  end
```
Will this query run every time `course.json` is requested? I have a concern about performance here. Some courses can have a large number of articles (potentially around 100K), so it would be nice to make sure this runs efficiently enough.

satrijo commented on app/views/courses/course.json.jbuilder:
```diff
@@ -54,6 +54,10 @@ json.course do
+  json.view_count_created number_to_human @course.view_sum_created
+  # Raw numbers for breakdown calculation
+  json.view_sum_raw @course.view_sum
+  json.view_sum_created_raw @course.view_sum_created
```
You were right, the JSON was confusing. I've simplified this based on your suggestion:

I removed the _raw value fields (view_sum_raw, view_sum_created_raw) from course.json.jbuilder.

I added the new pre-calculated, human-readable value view_count_existing directly from the backend.

The frontend (overview_stats.jsx) is now much simpler. It just displays the three values (view_count, view_count_created, view_count_existing) passed from the API and does no formatting or math for the tooltip.

satrijo commented:
> @satrijo please fix the linting and once CI is green, I'll review again

Hi, please check again. Thanks

gabina commented on app/assets/javascripts/components/overview/overview_stats.jsx:
```diff
@@ -13,6 +13,34 @@ const OverviewStats = ({ course }) => {
+    // If breakdown data is not available, fall back to simple doc
+    if (!hasBreakdown) {
+      return I18n.t(`metrics.${ArticleUtils.projectSuffix(courseData.home_wiki.project, 'view_count_doc')}`);
+    }
```
It seems like a strange design decision that in some cases you see a doc explaining how views are calculated, while in others you see only a breakdown without the explanation. However, as issue #6548 indicates, the view count doc is outdated, so maybe it’s fine not to show it.

satrijo replied:
You're right that this is inconsistent. The current implementation shows:

- **With breakdown**: Only the breakdown (no explanation of how views are calculated)
- **Without breakdown**: Only the doc (explanation of how views are calculated)

As you noted, issue #6548 indicates the view count doc is outdated. The breakdown provides more useful information (showing the split between new and existing articles), so showing only the breakdown when available seems reasonable. However, if you'd prefer, we could show both the breakdown AND the doc when breakdown is available.

gabina commented on app/models/course.rb:
```diff
@@ -632,6 +652,24 @@ def ensure_home_wiki_in_courses_wikis
+        )
+      )
+  end
+
```
Is this query the same as the one in `update_view_sum_created_based_on_first_revision`? If so, we should avoid the duplication.

satrijo replied:
I've refactored to eliminate the duplication:

- Created a shared class method `CourseCacheManager.calculate_view_sum_created_for_course(course)` that contains the query logic
- Both `CourseCacheManager#update_view_sum_created_based_on_first_revision` and the removed `Course#calculate_view_sum_created` now use this shared method
- This ensures consistency and makes maintenance easier

**Files changed:**

- `lib/course_cache_manager.rb`: Added shared class method `calculate_view_sum_created_for_course`
- `app/models/course.rb`: Removed duplicate `calculate_view_sum_created` method, now uses the shared method via fallback logic

satrijo commented on app/assets/javascripts/components/overview/overview_stats.jsx:
```diff
@@ -13,6 +13,34 @@ const OverviewStats = ({ course }) => {
+      && courseData.view_count_existing !== undefined
+      && courseData.view_count_existing !== null);
+
+    // If breakdown data is not available, fall back to simple doc
```
Breakdown data (`view_count_created and view_count_existing`) is not available in the following cases:

- Historical courses that haven't been updated since the migration: Courses that existed before the view_sum_created column was added and haven't been updated since then will have view_sum_created = 0 (the default). For these courses, we intentionally don't calculate the breakdown on-the-fly to:

-- Avoid performance issues (courses can have 100K+ articles)

-- Maintain consistency (since view_sum is a static cached value, mixing it with a dynamically calculated view_sum_created would be inconsistent)

- Courses with no new articles: If a course truly has no new articles created by participants, view_sum_created will be 0, and the breakdown would show "0 from new articles" and all views "from existing articles". However, we still show the breakdown in this case.

- Data inconsistency edge case: If view_sum_created somehow exceeds view_sum (which shouldn't happen but could due to timing issues or data inconsistencies), we skip the breakdown to avoid showing negative values for "existing articles".

The frontend checks for undefined or null values, which will occur when the backend doesn't include these fields in the JSON response (see `course.json.jbuilder `lines 59-62).

satrijo commented on app/models/course.rb:
```diff
@@ -614,6 +615,25 @@ def update_cache_from_timeslices
+    else
+      cached_value
+    end
+  end
```
You're absolutely right about the performance concerns. I've implemented a balanced approach:

- **Historical courses that haven't been updated**: Return the cached value (which defaults to 0) without on-the-fly calculation to avoid performance issues
- **Courses that have been updated but `view_sum_created` is still 0**: Calculate on-the-fly only if the course has been updated (`was_course_ever_updated?`) and has new articles. This handles courses that were updated before the `view_sum_created` column was added
- **Courses with recent updates**: Show the breakdown using cached values that were calculated during the update cycle

This approach:

- Avoids performance issues for historical courses that haven't been updated
- Provides accurate breakdowns for courses that have been updated but `view_sum_created` wasn't populated during the first update after migration
- Maintains consistency between `view_sum` and `view_sum_created` for courses that have been properly updated
